### PR TITLE
fix(dht): Recursive operation second path close condition

### PR DESCRIPTION
**Pull request by juslesan**
## Summary

In cases where the second routing path for recursive operations was no longer finding closer peers than the first path the RecursiveOperationSession could easily enter timeout state. this is because routing to the closest peer is not possible twice as it rejects the second message as a duplicate. To counter this if the second path finds the same closest node as the first path the operation will resolve.

This change also speeds up closing RecursiveFindSession with the FETCH_DATA mode when no data is behind the key

### CODE REVIEW

File: packages/dht/src/dht/recursive-operation/RecursiveOperationSession.ts [hunk 1]
```diff
@@ -45,6 +46,7 @@ export class RecursiveOperationSession extends EventEmitter<RecursiveOperationSe
     private timeoutTask?: NodeJS.Timeout 
     private completionEventEmitted = false
     private noCloserNodesReceivedCounter = 0
+    private noCloserNodesReceivedFrom: Set<NodeID> = new Set()
```
Comment (harbu) [inline packages/dht/src/dht/recursive-operation/RecursiveOperationSession.ts]: This set is never read from (only written to)?

File: packages/dht/test/integration/StoreOnDhtWithThreeNodes.test.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        await node1.stop()
+        await node2.stop()
+        simulator?.stop()
+    })
+
+    it('Node can store on three peer DHT', async () => {
+        const storedData1 = createMockDataEntry()
+        const storedData2 = createMockDataEntry()
+        await node1.getDataFromDht(storedData1.key)
+        await node2.getDataFromDht(storedData1.key)
```
Comment (harbu) [inline packages/dht/test/integration/StoreOnDhtWithThreeNodes.test.ts]: Why do we `getDataFromDht` before `storeDataToDht`? And why is the same not done to `entryPoint`?

**Review by harbu — CHANGES_REQUESTED** [review]

File: packages/dht/test/integration/StoreOnDhtWithThreeNodes.test.ts (hunk 2 above)
Comment (juslesan) [reply packages/dht/test/integration/StoreOnDhtWithThreeNodes.test.ts]: Before this PR fetching null data from the DHT would take a lot longer in smaller networks. Without the changes the test would timeout on these first calls

File: packages/dht/src/dht/recursive-operation/RecursiveOperationSession.ts (hunk 1 above)
Comment (harbu) [reply packages/dht/src/dht/recursive-operation/RecursiveOperationSession.ts]: What I meant with this comment is that the set is only `add`:ed to but never read or checked anywhere in the code?

File: packages/dht/test/integration/StoreOnDhtWithThreeNodes.test.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        await node1.stop()
+        await node2.stop()
+        simulator?.stop()
+    })
+
+    it('Node can store on three peer DHT', async () => {
+        const storedData1 = createMockDataEntry()
+        const storedData2 = createMockDataEntry()
+        await node1.getDataFromDht(getDhtAddressFromRaw(storedData1.key))
+        await node2.getDataFromDht(getDhtAddressFromRaw(storedData1.key))
```
Comment (harbu) [inline packages/dht/test/integration/StoreOnDhtWithThreeNodes.test.ts:45]:
```suggestion
        // Here we effectively test that fetching "null" data doesn't take too long. A test timeout here indicates an issue. 
        await node1.getDataFromDht(getDhtAddressFromRaw(storedData1.key))
        await node2.getDataFromDht(getDhtAddressFromRaw(storedData1.key))
        
```

**Review by harbu — COMMENTED** [review]
I would add a comment here for clarity.

**Review by harbu — APPROVED** [review]

**Pull request merged**